oeps/oep-0030-arch-pii-markup-and-auditing.rst

```diff
@@ -0,0 +1,269 @@
+-----------------------------
+In order to assist developers in remembering to check all new data for PII, each Open edX repository that might store such data will have a GitHub pull request template that reminds the developer and reviewers to check for the addition of such data in their commits and asks them to affirmatively state that either no such data exists or that it does exist and that appropriate retirement steps are, or will be, ready to retire that data before the request is merged.
+
+Docstring Annotations
+---------------------
+**Python**
+
+When adding or modifying **any** data storing models (ex. Django model, MongoDB model) an annotation must be added stating whether the model does or does not store PII. The annotation should be added to the comments of the storage class where the data storage is defined, or the calling method / function if there is no storage class. Calls to third-party services that store data must be annotated to indicate the presence of PII.
+
+It is important to note that under this OEP all Django model classes must be annotated with an assertion of PII / no PII to enable enforcement (see `Enforcement Tooling`_).
+
+These annotation should take the form of a Sphinx-style docstring:
+
+``.. pii:: <description of the PII>. Retired: <location / process of retirement>``
+
+Or, if there is no PII:
+
+``.. pii:: No PII``
+
+Example 1::
+
+    class ApiAccessRequest(TimeStampedModel):
+        """
+        Model to track API access for a user.
+
+        .. pii:: Contains website and employer information about a linked User. Retired: Directly in the LMSAccountRetirementView endpoint.
+        """
+
+If a project requires another project that stores PII, such as Django being used in edx-platform, the developer must annotate the place(s) in code where that package is being called to store the PII with the same docstring annotation as if it were a storage class.
+
+Example 2::
+
+    # ..pii:: Learner email is sent to Segment in the following line and will be associated with analytics data. Retired: In Segment retirement step in Tubular.
+
+The goal of this is to allow creation of `Documentation Tooling`_ which will automatically create documentation listing all of the known locations of PII in each repository.
+
+**Javascript**
+
+When adding in Javascript that results in storage of PII to a location that is not covered by other annotations (ex. Segment), annotations should be added to the location(s) in script where the data is being sent. The annotations should take the same form as in Python as Sphinx can also operate on Javascript for documentation.
+
+Example 1::
+
+    % if settings.LMS_SEGMENT_KEY:
+        <!-- begin segment footer -->
+        <!-- .. pii:: The user is identified to Segment by username and email here, analytics data will be associated with this user based on their browsing. See Segment documentation for details. Retired: In Segment retirement step in Tubular -->
```

Comment:
Has the team considered expanding on the annotation somewhat so the tooling can more easily parse out the PII data?  For example, instead of listing `username and email` within an English sentence, we could additionally list them in a separate `.. pii_data::` line right below.  This has the benefit of adding more **consistent** structure around the annotation and allows for better **discovery** - for example, finding all code that touches a user's email address.

Reply:
More structure may also require developers to specify an explicit retirement strategy of the PII: deleted, anonymized, retained.

Reply:
We discussed this in the arch review meeting, and updates for this should be in the next version which I hope to push today.